packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -1079,7 +1080,7 @@ class TextPainter {
         graphemeClusterLength *= 2;
         continue;
       }
-      final TextBox box = boxes.first;
+      final TextBox box = boxes.last;
```

Comment:
Interesting question. I'd be surprised if that was the behavior, but I can't immediately rule it out — in particular the dartdoc on getBoxesForRange doesn't say anything about the order of the boxes in the list.

... OK, and I wrote a test for this, and indeed the boxes seem to be in display order! Thanks for the catch.

I'll revise so that the change doesn't break anything that was working before. I expect it still won't work reliably in the presence of bidi, but fixing that may require bigger changes to this code.

Reply:
Done; PTAL.

I also filed #123424 to track that this logic doesn't seem like it can work in the bidi case. That's an issue this PR doesn't fix, and I think is best left out of scope because it will probably require bigger changes.